server/finder/handler/finder_handler.go

```diff
@@ -56,40 +57,60 @@ func (h *FinderHandler) Find(mhashes []multihash.Multihash) (*model.FindResponse
 
 		provResults := make([]model.ProviderResult, 0, len(values))
 		for j := range values {
-			provID := values[j].ProviderID
-			// Lookup provider info for each unique provider, look in local map
-			// before going to registry.
-			addrs, ok := provAddrs[provID]
+			iVal := values[j]
+			provID := iVal.ProviderID
+			addrs := h.fetchProviderAddrs(provID, iVal.ContextID, provAddrs, true)
+
+			if addrs == nil {
+				continue
+			}
+
+			// Check if there is an extended provider record for that provID
+			epRecord, ok := extendedProviders[provID]
 			if !ok {
-				pinfo, allowed := h.registry.ProviderInfo(provID)
-				if pinfo == nil {
-					// If provider not in registry, then provider was deleted.
-					// Tell the indexed core to delete the contextID for the
-					// deleted provider. Delete the contextID from the core,
-					// because there is no way to delete all records for the
-					// provider without a scan of the entire core valuestore.
-					go func(value indexer.Value) {
-						err := h.indexer.RemoveProviderContext(value.ProviderID, value.ContextID)
-						if err != nil {
-							log.Errorw("Error removing provider context", "err", err)
-						}
-					}(values[j])
-					// If provider not in registry, do not return in result.
-					continue
-				}
-				// Omit provider info if not allowed or marked as inactive.
-				if !allowed || pinfo.Inactive() {
-					continue
-				}
-				addrs = pinfo.AddrInfo.Addrs
-				provAddrs[provID] = addrs
+				epRecord = h.registry.ExtendedProviders(provID)
+				extendedProviders[provID] = epRecord
 			}
 
-			provResult, err := providerResultFromValue(values[j], addrs)
+			// Adding the main provider
+			provResult, err := providerResultFromValue(provID, iVal.ContextID, iVal.MetadataBytes, addrs)
 			if err != nil {
 				return nil, err
 			}
 			provResults = append(provResults, provResult)
+
+			if epRecord == nil {
+				continue
+			}
+
+			// If override is set to true at the context level then the chain level EPs should be ignored for this context ID
+			override := false
+
+			// Adding context-level EPs if they exist
+			if contextualEpRecord, ok := epRecord.ContextualProviders[string(iVal.ContextID)]; ok {
+				override = contextualEpRecord.Override
+				for _, epInfo := range contextualEpRecord.Providers {
+					provResult, err := createExtendedProviderResult(epInfo, iVal)
+					if err != nil {
+						return nil, err
```

Comment:
If there is an error creating the `provResutl` I assume that means the data is unusable. Maybe it should be removed?  Also, should the bad data be skipped here to return what good data there is?

Reply:
Good catch. I revisited `createExtendedProviderResult` and it actually didn't need to return `error` at all (it was always `nil`). Removed the error checking logic here.

Reply:
Awesome! Most efficient code is no code.